### PR TITLE
Fixes, new commands and test for memkind frag

### PR DESCRIPTION
**Pull request by abonisl**
Populate & defrag fixes, new shell debug commands for frag ratio and new test (based on original memefficiency) for memkind defrag (its not finished yet)

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/redis/69)

### CODE REVIEW

**Comment by mbiesek** [comment]
This section should be updated withdescription of new commands, ( frag and fragM ) https://github.com/pmem/redis/blob/d59995ec2867be18a3fcf5e8b721b9221ef8c97c/src/debug.c#L270

File: src/debug.c [hunk 1]
```diff
@@ -556,6 +559,18 @@ void debugCommand(client *c) {
         changeReplicationId();
         clearReplicationId2();
         addReply(c,shared.ok);
+	} else if (!strcasecmp(c->argv[1]->ptr,"frag")) {
+		sds d = sdsempty();
+		size_t frag_bytes;
```
Comment (mbiesek) [inline src/debug.c]: Cosmetics but please be consitent here with spaces

File: src/debug.c [hunk 2]
```diff
@@ -49,6 +49,9 @@
 #endif
 #endif
 
+float getAllocatorFragmentation(size_t *out_frag_bytes);
```
Comment (mbiesek) [inline src/debug.c]: IMHO it should be in server.h under debugging section: https://github.com/pmem/redis/blob/dcbc3a10d4ee384d6d6c188fd19269e391f89736/src/server.h#L2116

**Comment by KFilipek** [comment]
Please provide more detailed description if changes are bigger than few lines. 
What fixes are made, command and use. We should confront ideas with implementation.

**Comment by abonisl** [comment]
*[src/debug.c, line 52 at r1](https://reviewable.io/reviews/pmem/redis/69#-LFaOA_wuA4_h8Dczbro-r1-52:-LFphJCw-HLd6Kqa2moB:bb6un1j) ([raw file](https://github.com/pmem/redis/blob/d021cb9a78121cc0974c1c34da64d9f800f8b608/src/debug.c#L52)):*
<details><summary><i>Previously, mbiesek wrote…</i></summary><blockquote>

IMHO it should be in server.h under debugging section: https://github.com/pmem/redis/blob/dcbc3a10d4ee384d6d6c188fd19269e391f89736/src/server.h#L2116
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/redis/69)*

**Comment by abonisl** [comment]
*[src/debug.c, line 564 at r1](https://reviewable.io/reviews/pmem/redis/69#-LFaOA_wuA4_h8Dczbro-r1-564:-LFphSfV66KJB1HVo--5:b-886gsx) ([raw file](https://github.com/pmem/redis/blob/d021cb9a78121cc0974c1c34da64d9f800f8b608/src/debug.c#L564)):*
<details><summary><i>Previously, mbiesek wrote…</i></summary><blockquote>

Cosmetics but please be consitent here with spaces 
</blockquote></details>

Fixed

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/redis/69)*

**Comment by abonisl** [comment]
@KFilipek  they are super small and was discussed at the last meeting
populate function need to use createStringObjectM() instead of createStringObject,
activeDefragSdsM() should run activeDefragSdsA with m_alloc
deleted setting server.keys_on_pm to false in InitServer()

+ new commands:
debug frag - show frag level for jemalloc
debug fragM - show frag level for memkind (second jemalloc)

We need them to check frag level in tests, also there was a TODO section in previous test:

> # TODO: we need to expose more accurate fragmentation info

so its done now :)

**Comment by abonisl** [comment]
Thanks to @jschmieg  very important fix for memkind defrag in latest commit. Previously memkind allocations were moved to ram.
+ small changes for memkind defrag test, now it works great

**Comment by michalbiesek** [comment]
Review status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @mbiesek, @abonisl, and @KFilipek)

---

*[tests/unit/memefficiencym.tcl, line 73 at r3](https://reviewable.io/reviews/pmem/redis/69#-LFqwFx52TBnjhNRwtT_:-LFqwFx64sX0tJ737G9L:b-r3iyah) ([raw file](https://github.com/pmem/redis/blob/ebf9f19efc901a3b216cb95e95da51c470d38174/tests/unit/memefficiencym.tcl#L73)):*
> ```Tcl
>                 }
> 				
>                 # TODO: we need to expose more accurate fragmentation info
> ```

This comment shouldn't be described  with "TODO" and "Update"- it will be seen in repository as one change so there is no need to use TODO and Update tags. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/redis/69)*

**Comment by michalbiesek** [comment]
*[tests/unit/memefficiencym.tcl, line 88 at r3](https://reviewable.io/reviews/pmem/redis/69#-LFqx9tLEOZEhBAIIIkf:-LFqx9tLEOZEhBAIIIkg:bf1ktzs) ([raw file](https://github.com/pmem/redis/blob/ebf9f19efc901a3b216cb95e95da51c470d38174/tests/unit/memefficiencym.tcl#L88)):*
> ```Tcl
>                 set misses2 [s active_defrag_misses]
>                 assert {$misses2 == $misses}
> 				
> ```

replace tabs with spaces.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/redis/69)*

**Comment by michalbiesek** [comment]
*[src/debug.c, line 52 at r3](https://reviewable.io/reviews/pmem/redis/69#-LFqxJww6zaurZCAD4Y_:-LFqxJww6zaurZCAD4Ya:bvag2pn) ([raw file](https://github.com/pmem/redis/blob/ebf9f19efc901a3b216cb95e95da51c470d38174/src/debug.c#L52)):*
> ```C
> #endif
> ```

Remove this change

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/redis/69)*

**Comment by michalbiesek** [comment]
Consider clear ( squash) commit history, for example commit with title "cosmetic changes" is not necessary

**Comment by abonisl** [comment]
Ok, fixed @michalbiesek 
I will try to use git squash now

**Comment by michalbiesek** [comment]
Reviewed 1 of 2 files at r2, 1 of 2 files at r3, 2 of 2 files at r4.
Review status: 4 of 5 files reviewed, 2 unresolved discussions (waiting on @mbiesek and @KFilipek)

---



*Comments from [Reviewable](https://reviewable.io/reviews/pmem/redis/69)*

**Comment by michalbiesek** [comment]
:lgtm:

---

Review status: 4 of 5 files reviewed, 2 unresolved discussions (waiting on @mbiesek and @KFilipek)

---



*Comments from [Reviewable](https://reviewable.io/reviews/pmem/redis/69#-:-LFrLAQd1c10wk0lpyL_:bnfp4nl)*

**Comment by karpiop** [comment]
:lgtm:

---

Reviewed 1 of 4 files at r1.
Review status: all files reviewed, 2 unresolved discussions (waiting on @mbiesek and @KFilipek)

---



*Comments from [Reviewable](https://reviewable.io/reviews/pmem/redis/69#-:-LFvxdnt0hcYgmuG7Utd:bnfp4nl)*

**Review by michalbiesek — APPROVED** [review]

**Comment by abonisl** [comment]
Added commit with latest fixes for m_alloc&z_alloc usage. All tests from set.tcl (and 'make test' too) passed with memkind jemalloc 4.5

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/69)* status: 2 of 7 files reviewed, 3 unresolved discussions (waiting on @michalbiesek, @mbiesek, @abonisl, and @KFilipek)

___
*[deps/memkind/src/memkind.c, line 719 at r6](https://reviewable.io/reviews/pmem/redis/69#-LLZVx0017s9bXhHaxq0:-LLZVx027u95VZZ0ehF7:b-2p8id4) ([raw file](https://github.com/pmem/redis/blob/bd5b508504841112ea6357386108f35a0cb5a201/deps/memkind/src/memkind.c#L719)):*
> ```
> int memkind_range(struct memkind **kind, void *ptr)
> ```
Change definition to:
-int memkind_range(struct memkind *kind, void *ptr)

Additionally change in this function :
from struct memkind_pmem *priv = (*kind)->priv; 
to struct memkind_pmem *priv = kind->priv;

Change calling function:
from serverAssert(memkind_range(&server.pmem_kind1, ptr)); 
to serverAssert(memkind_range(server.pmem_kind1, ptr));

**Review by michalbiesek — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 8 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/69#-:-LMaoI2w6eNRcvE-fq_Z:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @mbiesek and @KFilipek)

**Comment by abonisl** [comment]
Geo problem fixed, should be merged with pr #76

**Review by KFilipek — COMMENTED** [review]
> Reviewed 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/69)* status: all files reviewed, 2 unresolved discussions (waiting on @mbiesek)

**Review by KFilipek — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r4, 3 of 8 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/69#-:-LP_2ftJ-2Q2Qfxquu4M:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @mbiesek)

**Pull request merged**